File: _tests/evidence/v180-apply/apply-tokens.py
#!/usr/bin/env python3
"""
Surgical --od-* token VALUE replacement (Helix v180-apply).

Applies the brand-anchored hybrid candidate token VALUES onto the live MACHINA
brand CSS, block-for-block, WITHOUT touching component/effect CSS, comments,
ordering, or the --vd-* MACHINA knobs.

For each of the three token blocks the candidate defines
  (1) :root { }
  (2) :root[data-theme="dark"] { }
  (3) @media (prefers-color-scheme:dark){ :root:not([data-theme="light"]) { } }
we:
  - replace the VALUE of every --od-* token that both the candidate block and
    the matching live block define (value = text between ':' and ';'; any
    trailing inline comment after ';' is preserved byte-for-byte),
  - APPEND (before the block's closing brace, at the block's own indent) any
    --od-* token the candidate block defines that the live block lacks,
  - leave every other declaration (incl. --vd-*, --od-shadow-lg overrides the
    candidate doesn't define, etc.) untouched.

Deterministic, idempotent, no third-party deps. Prints a full audit.
"""
import re
import sys

CAND = "/Volumes/T7/Projects/vasic/design-toolkit/proposed/vasic-digital.od-tokens.css"
LIVE = "/Volumes/T7/Projects/vasic/design-system/brand-vasic-digital/vasic-digital.css"

# (name, header-regex) for each block, in cascade order.
BLOCKS = [
    ("root",  re.compile(r'(?m)^:root\s*\{')),
    ("dark",  re.compile(r'(?m)^:root\[data-theme="dark"\]\s*\{')),
    ("media", re.compile(r'(?m)^[ \t]*:root:not\(\[data-theme="light"\]\)\s*\{')),
]

DECL = re.compile(r'--([A-Za-z0-9-]+)\s*:\s*([^;]*);')


def find_block(text, header_re):
    """Return (start_of_body, end_of_body_excl_brace, body_text) for the
    brace-matched block whose opening brace follows the header match."""
    m = header_re.search(text)
    if not m:
        raise SystemExit(f"header not found: {header_re.pattern}")
    brace = text.index('{', m.start())
    depth = 0
    i = brace
    while i < len(text):
        c = text[i]
        if c == '{':
            depth += 1
        elif c == '}':
            depth -= 1
            if depth == 0:
                return brace + 1, i, text[brace + 1:i]
        i += 1
    raise SystemExit("unbalanced braces")


def parse_od_tokens(body):
    """Ordered dict of --od-* tokens -> value, as defined in a block body."""
    out = {}
    for name, val in DECL.findall(body):
        if name.startswith("od-"):
            out["--" + name] = val.strip()
    return out


def block_indent(body):
    """Detect the leading-whitespace indent of declarations in this block."""
    m = re.search(r'(?m)^([ \t]+)--od-', body)
    return m.group(1) if m else "  "


def main():
    with open(CAND, encoding="utf-8") as f:
        cand_text = f.read()
    with open(LIVE, encoding="utf-8") as f:
        live_text = f.read()

    # candidate token maps per block
    cand_maps = {}
    for name, hre in BLOCKS:
        _, _, body = find_block(cand_text, hre)
        cand_maps[name] = parse_od_tokens(body)

    total_changed = 0
    total_same = 0
    total_added = 0
    audit = []

    # Rebuild the live text block by block. Because appends change offsets, we
    # process blocks left-to-right and re-locate each block in the mutated text.
    for name, hre in BLOCKS:
        b_start, b_end, body = find_block(live_text, hre)
        live_tokens = parse_od_tokens(body)
        indent = block_indent(body)
        new_body = body

        for tok, newval in cand_maps[name].items():
            if tok in live_tokens:
                oldval = live_tokens[tok]
                if oldval == newval:
                    total_same += 1
                    continue
                # replace value between ':' and ';' for THIS token only (1st hit)
                pat = re.compile(r'(' + re.escape(tok) + r'\s*:\s*)([^;]*)(;)')
                def _sub(mm):
                    return mm.group(1) + newval + mm.group(3)
                new_body, n = pat.subn(_sub, new_body, count=1)
                if n != 1:
                    raise SystemExit(f"[{name}] failed to replace {tok}")
                total_changed += 1
                audit.append(f"  [{name}] {tok}: {oldval}  ->  {newval}")

        # append tokens present in candidate block but missing from live block
        missing = [t for t in cand_maps[name] if t not in live_tokens]
        if missing:
            add_lines = "".join(
                f"{indent}{t}: {cand_maps[name][t]};\n" for t in missing
            )
            # insert right before the block's closing brace, preserving the
            # newline+indent that precedes the '}'.
            if not new_body.endswith("\n"):
                new_body = new_body + "\n"
            new_body = new_body + add_lines
            for t in missing:
                total_added += 1
                audit.append(f"  [{name}] +ADD {t}: {cand_maps[name][t]}")

        live_text = live_text[:b_start] + new_body + live_text[b_end:]

    with open(LIVE, "w", encoding="utf-8") as f:
        f.write(live_text)

    print("=== v180-apply token audit ===")
    for line in audit:
        print(line)
    print("-------------------------------")
    print(f"values changed : {total_changed}")
    print(f"already equal  : {total_same}")
    print(f"tokens added   : {total_added}")


if __name__ == "__main__":
    main()

File: _tests/evidence/v180-apply/apply-tokens.mjs
#!/usr/bin/env node
// Surgical token-VALUE replacement — apply the crimson-anchored hybrid candidate
// to the LIVE milosvasic brand CSS while PRESERVING the TERMINAL BRUTALIST design
// (Anton display) and matching the approved STAGE-mv preview (== current live look).
//
// WHAT GETS THE CANDIDATE VALUE (in the three BASE blocks):
//   accent ramp (--od-accent-50..900), neutrals (bg/surface/surface-2/text/
//   text-muted/border — NEUTRALIZED to warm-neutral, de-greened), accent
//   semantics (accent/hover/active/on-accent/focus), status (success/warning/
//   danger/shadow-color), radius (--od-radius-*), shadow recipes (--od-shadow-*).
//
// WHAT IS PRESERVED AT LIVE VALUES (NOT replaced) — these are the brutalist
// design system's rhythm/identity and are NOT re-declared by the trailing
// brutalist :root override, so replacing them would break the layout / identity:
//   fonts (--od-font-*), type-scale (--od-fs-*), spacing (--od-space-*),
//   line-height (--od-lh-*), tracking (--od-tracking-*), motion (--od-dur-*,
//   --od-ease-*), z-index (--od-z-*), --od-container-max.
//   (Proof: replacing --od-fs-* / --od-space-* blows up the hero lede & spacing,
//    35% pixel diff vs STAGE — see evidence/v180-apply.)
//
// BRUTALIST OVERRIDE (2nd :root): only --od-accent-700 is repointed to the
// candidate crimson #93474f so the *computed* accent ramp is crimson-anchored
// (verification gate). The primary --od-accent stays #a31e39 (STAGE-exact), and
// Anton / sharp radius / hard print-shadows / warm-cream neutrals are untouched.
//
// Usage: node apply-tokens.mjs <candidate.css> <live.css> <out.css>
import { readFileSync, writeFileSync } from 'node:fs';

const [candPath, livePath, outPath] = process.argv.slice(2);
if (!candPath || !livePath || !outPath) {
  console.error('usage: apply-tokens.mjs <candidate.css> <live.css> <out.css>');
  process.exit(2);
}
const cand = readFileSync(candPath, 'utf8');
const live = readFileSync(livePath, 'utf8');

// Tokens PRESERVED at live values (skip replacement in base blocks).
const PRESERVE = (name) =>
  name.startsWith('--od-font-') ||
  name.startsWith('--od-fs-') ||
  name.startsWith('--od-space-') ||
  name.startsWith('--od-lh-') ||
  name.startsWith('--od-tracking-') ||
  name.startsWith('--od-dur-') ||
  name.startsWith('--od-ease-') ||
  name.startsWith('--od-z-') ||
  name === '--od-container-max';

// Neutralized (de-greened, warm-neutral) surface/neutral values — replace the
// candidate's faintly-green MCU-Expressive neutrals. Accent ramp untouched.
const NEUTRALS_LIGHT = {
  '--od-bg': '#faf9f7', '--od-surface': '#f4f2ef', '--od-surface-2': '#eeebe7',
  '--od-text': '#1c1a17', '--od-text-muted': '#4a453f', '--od-border': '#c8c3bc',
};
const NEUTRALS_DARK = {
  '--od-bg': '#12100e', '--od-surface': '#1b1917', '--od-surface-2': '#211e1b',
  '--od-text': '#e7e3dd', '--od-text-muted': '#c3bcb2', '--od-border': '#453f38',
};

function blockBody(src, headerRe) {
  const m = headerRe.exec(src);
  if (!m) throw new Error('block not found: ' + headerRe);
  const open = src.indexOf('{', m.index);
  let depth = 0, i = open;
  for (; i < src.length; i++) {
    if (src[i] === '{') depth++;
    else if (src[i] === '}') { depth--; if (depth === 0) break; }
  }
  return { start: open + 1, end: i, body: src.slice(open + 1, i) };
}
function tokenMap(body) {
  const map = new Map();
  const re = /(--od-[a-z0-9-]+)\s*:\s*([^;]+);/gi; let m;
  while ((m = re.exec(body))) map.set(m[1], m[2].trim());
  return map;
}

const candRoot = tokenMap(blockBody(cand, /:root\s*\{/).body);
const candDark = tokenMap(blockBody(cand, /:root\[data-theme="dark"\]\s*\{/).body);
const candMediaOuter = blockBody(cand, /@media\s*\(prefers-color-scheme:\s*dark\)\s*\{/).body;
const candMedia = tokenMap(blockBody(candMediaOuter, /:root:not\(\[data-theme="light"\]\)\s*\{/).body);

// Apply neutral overrides onto candidate maps (so before/after is auditable).
for (const [k, v] of Object.entries(NEUTRALS_LIGHT)) if (candRoot.has(k)) candRoot.set(k, v);
for (const [k, v] of Object.entries(NEUTRALS_DARK)) { if (candDark.has(k)) candDark.set(k, v); if (candMedia.has(k)) candMedia.set(k, v); }

function rewriteBlock(src, headerRe, map, changed, { preserve = true } = {}) {
  const { start, end } = blockBody(src, headerRe);
  const before = src.slice(start, end);
  const after = before.replace(/(--od-[a-z0-9-]+)(\s*:\s*)([^;]+)(;)/gi, (full, name, sep, val, sc) => {
    if (preserve && PRESERVE(name)) return full;      // keep live rhythm/identity
    if (!map.has(name)) return full;
    const nv = map.get(name);
    if (val.trim() === nv) return full;
    changed.push({ name, from: val.trim(), to: nv });
    return name + sep + nv + sc;
  });
  return src.slice(0, start) + after + src.slice(end);
}

const changed = { root: [], dark: [], media: [], brutalist: [] };
let out = live;
out = rewriteBlock(out, /@media\s*\(prefers-color-scheme:\s*dark\)\s*\{[\s\S]*?:root:not\(\[data-theme="light"\]\)\s*\{/, candMedia, changed.media);
out = rewriteBlock(out, /:root\s*\{/, candRoot, changed.root);
out = rewriteBlock(out, /:root\[data-theme="dark"\]\s*\{/, candDark, changed.dark);

// Brutalist override: repoint ONLY --od-accent-700 to the candidate crimson so
// the computed accent ramp is crimson-anchored. Everything else in the brutalist
// blocks (Anton, radius, hard shadows, #a31e39 primary accent, warm neutrals) is
// left byte-for-byte. We patch the SECOND occurrence (the brutalist :root@683).
{
  const target = '--od-accent-700';
  const candVal = candRoot.get(target); // #93474f (accent ramp — not neutralized)
  // Find the brutalist light :root (the one that also sets --od-font-display:Anton)
  const bru = blockBody(out, /:root\{\n\s*--od-bg:#f4f1ea;/);
  const body = out.slice(bru.start, bru.end);
  const patched = body.replace(/(--od-accent-700\s*:\s*)([^;]+)(;)/i, (full, pfx, val, sc) => {
    if (val.trim() === candVal) return full;
    changed.brutalist.push({ name: target, from: val.trim(), to: candVal, block: 'brutalist :root' });
    return pfx + candVal + sc;
  });
  out = out.slice(0, bru.start) + patched + out.slice(bru.end);
}

writeFileSync(outPath, out);

const summarize = (label, arr) => {
  console.log(`\n[${label}] ${arr.length} changed`);
  for (const c of arr) console.log(`  ${c.name}: ${c.from}  ->  ${c.to}`);
};
summarize('base :root', changed.root);
summarize('base dark', changed.dark);
summarize('base @media dark', changed.media);
summarize('brutalist :root (accent anchor only)', changed.brutalist);
console.log(`\nwrote ${outPath} (${out.length} bytes; live was ${live.length} bytes)`);
